jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java

```diff
@@ -0,0 +1,465 @@
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.thread.strategy;
+
+import java.io.Closeable;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.concurrent.Executor;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.annotation.ManagedOperation;
+import org.eclipse.jetty.util.component.ContainerLifeCycle;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.eclipse.jetty.util.thread.ExecutionStrategy;
+import org.eclipse.jetty.util.thread.Invocable;
+import org.eclipse.jetty.util.thread.TryExecutor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>An adaptive execution strategy that uses the {@link Invocable} status
+ * of both the task and the current thread to select an optimal strategy
+ * that prioritizes executing the task immediately in the current
+ * producing thread if it can be done so without thread starvation issues.</p>
+ *
+ * This strategy selects between the following sub-strategies:
+ * <dl>
+ *     <dt>ProduceConsume(PC)</dt>
+ *     <dd>The producing thread consumes the task by executing it directly and then
+ *     continues to produce.</dd>
+ *     <dt>ProduceInvokeConsume(PIC)</dt>
+ *     <dd>The producing thread consumes the task by invoking it with {@link Invocable#invokeNonBlocking(Runnable)}
+ *     and then continues to produce.</dd>
+ *     <dt>ProduceExecuteConsume(PEC)</dt>
+ *     <dd>The producing thread dispatches the task to a thread pool to be executed and then immediately resumes
+ *     producing.</dd>
+ *     <dt>ExecuteProduceConsume(EPC)</dt>
+ *     <dd>The producing thread consumes the task by executing it directly (as in PC mode) but then races with
+ *     a pending producer thread to take over production.
+ *     </dd>
+ * </dl>
+ * The sub-strategy is selected as follows:
+ * <dl>
+ *     <dt>PC</dt><dd>If the produced task has been invoked with {@link Invocable#invokeNonBlocking(Runnable)}
+ *     to indicate that it is {@link Invocable.InvocationType#NON_BLOCKING}.</dd>
+ *     <dt>EPC</dt><dd>If the producing thread is not {@link Invocable.InvocationType#NON_BLOCKING}
+ *     and a pending producer thread is available, either because there is already a pending producer
+ *     or one is successfully started with {@link TryExecutor#tryExecute(Runnable)}.</dd>
```

Comment:
```suggestion
 *     <dt>EPC</dt>
 *    <dd>When the producing thread is not marked as non-blocking, and a producer thread is available,
 *    since {@link TryExecutor#tryExecute(Runnable)} returned {@code true}.</dd>
```

Reply:
ditto